R/reindent.R

```diff
@@ -0,0 +1,134 @@
+  }
+  pd_nested
+}
+
+#' @describeIn update_indention_ref Updates the reference id for all
+#'   tokens in `pd_nested` if `pd_nested` contains a function declaration.
+#'   Tokens inside a function declaration are are re-indented,
+#'   that is, they are indented up to the level at which the token FUNCTION
+#'   ends in terms of col2.
+#' @examples
+#' \dontrun{
+#' a <- function(x,
+#'               y) {
+#' x + y
+#' }
+#' }
+update_indention_ref_fun_dec <- function(pd_nested) {
+  if (pd_nested$token[1] == c("FUNCTION") &&
+      nrow(pd_nested) > 3) {
+    seq <- 3:(nrow(pd_nested) - 1)
+    pd_nested$indention_ref_id[seq] <- pd_nested$id[1]
+  }
+  pd_nested
+}
+
+#' Check whether a parse table corresponds to a a certain expression
+#'
+#' @param pd A parse table.
+#' @name pd_is
+NULL
+
+#' @describeIn pd_is Checks whether `pd` contains an expression wrapped in
+#'   curley brackets.
+is_curly_expr <- function(pd) {
+  if (is.null(pd)) return(FALSE)
+  pd$token[1] == "'{'"
+}
+
+#' @describeIn pd_is Checks whether `pd` is a function call.
+is_function_call <- function(pd) {
+  if (is.null(pd)) return(FALSE)
+  if (is.na(pd$token_before[2])) return(FALSE)
+  pd$token_before[2] == "SYMBOL_FUNCTION_CALL"
+}
+
+
+#' Apply reference indention to tokens
+#'
+#' Applies the reference indention created with functions
+#' [update_indention_ref()] to the flattened parse table. The indention
+#' is applied to all token that inherit from a reference token sequentially,
+#' i.e. by looping over the target tokens.
+#' @inheritParams apply_ref_indention_one
+apply_ref_indention <- function(flattened_pd) {
+  target_tokens <- which(flattened_pd$id %in% flattened_pd$indention_ref_id)
+
+  # needs sequential update
+  for (target_token in target_tokens) {
+    flattened_pd <- apply_ref_indention_one(flattened_pd, target_token)
+  }
+  flattened_pd
+}
+
+#' Applying reference indention of a target token
+#'
+#' Applies the indention level of `target_token` to all tokens that have
+#' `target_token` as a reference. This includes adding spaces to the first
+#' tokens on a line and updating the column `col1` and `col2` for all token
+#' on that line so they are kept updated.
+#' @param flattened_pd A flattened parse table
+#' @param target_token The index of the token from wich the indention level
+#'   should be applied to other tokens.
+apply_ref_indention_one <- function(flattened_pd, target_token) {
+  token_points_to_ref <-
+    flattened_pd$indention_ref_id == flattened_pd$id[target_token]
+  first_token_on_line <- flattened_pd$lag_newlines > 0L
+  token_to_update <- which(token_points_to_ref & first_token_on_line)
+
+  # udate spaces
+  copied_spaces <- flattened_pd$col2[target_token] + 1
+  old_spaces <- flattened_pd$lag_spaces[token_to_update[1]]
+  shift <- copied_spaces - old_spaces
+  flattened_pd$lag_spaces[token_to_update] <-
+    flattened_pd$lag_spaces[token_to_update] + shift
+
+  # update col1 / col2
```

Comment:
I wonder if we really need to drag along `col1` and `col2` (#112), but we can scrap that later.